### PR TITLE
Fix renaming the old_relation in Oracle 12c+ when using DBT table materialization with missing database name in profiles.yml

### LINKED ISSUE #4: [Bug] Renaming the old_relation does not work in Oracle 12c+ when using DBT table materialization with missing database name in profiles.yml
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```
1 of 1 START table model staging.test.......................................................................... [RUN]
1 of 1 ERROR creating table model staging.test................................................................. [ERROR in 0.96s]
oracle adapter: Oracle error: ORA-00955: name is already used by an existing object

Finished running 1 table models in 0.96s.

Completed with 1 errors and 0 warnings:

Database Error in model test (models/generated/airbyte_tables/staging/test.sql)
  ORA-00955: name is already used by an existing object
  compiled SQL at ../build/run/airbyte_utils/models/generated/airbyte_tables/staging/test.sql

Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1
```

### Expected Behavior

```1 of 1 START table model staging.test.......................................................................... [RUN]
1 of 1 OK created table model staging.test..................................................................... [OK in 1.45s]

Finished running 1 table models in 1.45s.

Completed successfully

Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

### Steps To Reproduce

Run the DBT table materialization twice.

### Relevant log output using `--debug` flag enabled

_No response_

### Environment

```markdown
- OS: Oracle Linux Server 8.5
- Python: 3.9.9
- dbt: 1.0.6
```

### What Oracle database version are you using dbt with?

Tested using
- Oracle Database 12c Enterprise Edition Release 12.2.0.1.0 - 64bit Production
**and**
- Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production

### Additional Context

Pull request fix #3

### PR DESCRIPTION
**Pull request by ThoSap**
Hi,

first of all, thank you very much for releasing an official DBT Oracle DB adapter/profile with an up-to-date DBT version!!! 😍
I tried this adapter today locally together with some Airbyte modifications and found that version 1.0.0 does not rename the already existing destination table to a backup table when using DBT table materialization and therefore I get the following error:

```python
1 of 1 ERROR creating table model staging.test................................................................. [ERROR in 0.96s]
2022-05-04 08:41:55 normalization > 08:41:55  oracle adapter: Oracle error: ORA-00955: name is already used by an existing object
```
With dbt run --debug
```python
Database Error in model test (models/generated/airbyte_tables/staging/test.sql)
2022-05-04 08:42:00 normalization > 08:41:55.566833 [error] [MainThread]:   ORA-00955: name is already used by an existing object
2022-05-04 08:42:00 normalization > 08:41:55.567157 [error] [MainThread]:   compiled SQL at ../build/run/airbyte_utils/models/generated/airbyte_tables/staging/test.sql
```

This is caused by the fact that the database name is passed to function `adapter.get_relation()` which should not be the case for Oracle DB.
https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/include/oracle/macros/materializations/table/table.sql#L22
In Oracle DB the database name is never a prefix when making a select as it does not even work.
✅ `schema.relation`
❌ `database.schema.relation`

Also at the DBT/Jinija macro `oracle__list_relations_without_caching` the filter for the table_catalog (database) should be removed as usually no one ever will pass the database name in the `profiles.yml`.

Fixes #4
Fixes techindicium/dbt-oracle#8
Fixes techindicium/dbt-oracle#32

### CODE REVIEW

**Comment by aosingh** [comment]
Thank you @ThoSap for testing and contributing. 

While we review to get this merged, could you sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/) ?

The bottom of your commit message must have the following line using your name
and e-mail address as it appears in the OCA Signatories list.

```
Signed-off-by: Your Name <you@example.org>
```

This can be automatically added to pull requests by committing with:

```
git commit --signoff
````

**Comment by ThoSap** [comment]
I just wanted to create an **Individual OCA** but the `oracle/dbt-oracle` repo is not available for selection and it is a required field.
Can you help @aosingh ?

**Comment by aosingh** [comment]
Sure, let me check and I will get back to you

**Comment by aosingh** [comment]
Hi @ThoSap 

Thanks once again for your help! 

I would like to better understand the problem and the fixes. The reason is, I am trying to reproduce the issue but it works as expected with the current version of dbt-oracle.

Test Model - https://github.com/oracle/dbt-oracle/blob/main/dbt_adbs_test_project/models/promotion_costs.sql
Oracle  Database version - 19c

I ran the model 3 times and each time it completed successfully. With `--debug` flag enabled i could also see the backup table being dropped using the below PL/SQL. 

```
DECLARE
     dne_942    EXCEPTION;
     PRAGMA EXCEPTION_INIT(dne_942, -942);
     attempted_ddl_on_in_use_GTT EXCEPTION;
     pragma EXCEPTION_INIT(attempted_ddl_on_in_use_GTT, -14452);
  BEGIN
     SAVEPOINT start_transaction;
     EXECUTE IMMEDIATE 'DROP table dbt_test.promotion_costs__dbt_backup cascade constraint';
     COMMIT;
  EXCEPTION
     WHEN attempted_ddl_on_in_use_GTT THEN
        NULL; -- if it its a global temporary table, leave it alone.
     WHEN dne_942 THEN
        NULL;
  END;

```

**Comment by ThoSap** [comment]
The thing with this change is that it would no longer be necessary to specify the `database` name (in **profiles.yml**) which normally is _never_ needed in any other application using OJDBC or cx_Oracle when connecting to an Oracle DB.
The _database name_ (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) is not required for OJDBC, cx_Oracle and not even in SQL Developer, but only the service name or SID instead.

I don't think the database name is needed at all as the service name should suffice with these PR changes, maybe we can eliminate this property altogether.
But it could be that I don't get the technical reason why the database name is really needed in DBT Oracle 🤔

This probably has historic reasons and also there is the confusion, where if one specifies the **profiles.yml** property `dbname` **but not** `database`, the property `dbname` with alias [database](https://github.com/oracle/dbt-oracle/blob/main/dbt/adapters/oracle/connections.py#L83) gets to be the service name and then the connection works anyways.
https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/connections.py#L129-L134
Example **profiles.yml**:
```yaml
config:
  partial_parse: true
  printer_width: 120
  send_anonymous_usage_stats: false
  use_colors: true
dbt_test:
  target: prod
  outputs:
    prod:
      type: oracle
      protocol: tcp
      host: yourhost.example.com
      port: 1521
      schema: test
      user: test
      pass: SECRET_PW
      # Connection wors anyways as in connections.py#L83 the dbname property gets alias database and if the service property is not specified the database internally is the service in the end
      # https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/connections.py#L134
      dbname: test.service.name.example.com
      threads: 4
```

For my use case using Airbyte (I plan to do there a PR also with the new oracle/dbt-oracle adapter), the `database` property in **profiles.yml** would pose a breaking change in the [Oracle Destination Connector](https://docs.airbyte.com/integrations/destinations/oracle/), as suddenly users need to specify the _database_ name (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) as a [mandatory field](https://github.com/airbytehq/airbyte/blob/v0.36.9-alpha/airbyte-integrations/connectors/destination-oracle/src/main/resources/spec.json) if they want to use the Oracle [Basic Normalization](https://docs.airbyte.com/understanding-airbyte/basic-normalization/) which is already fully implemented in Airbyte but currently not enabled due to the fact that the current Airbyte Oracle normalization implementation is still using DBT 0.19.1 from `techindicium/dbt-oracle:0.4.3`.

**Comment by ThoSap** [comment]
Also the `OracleIncludePolicy` for database is `False` anyways 😁
https://github.com/oracle/dbt-oracle/blob/1.0.0/dbt/adapters/oracle/relation.py#L32

**Review by aosingh — COMMENTED** [review]
Could we revert the change in  `dbt/adapters/oracle/impl.py` because in `OracleIncludePolicy` the property `database` is False. Additionally `IncludePolicy` is configurable in dbt_project.yml ?

**Comment by ThoSap** [comment]
> Could we revert the change in `dbt/adapters/oracle/impl.py` because in `OracleIncludePolicy` the property `database` is False. Additionally `IncludePolicy` is configurable in dbt_project.yml ?

Without this change it is not working, even though `OracleIncludePolicy` the property `database` is `False`.

**Comment by aosingh** [comment]
> > Could we revert the change in `dbt/adapters/oracle/impl.py` because in `OracleIncludePolicy` the property `database` is False. Additionally `IncludePolicy` is configurable in dbt_project.yml ?
> 
> Without this change it is not working, even though `OracleIncludePolicy` the property `database` is `False`.

Ok, I will check this and let you know.

**Review by aosingh — COMMENTED** [review]
I wanted to start by keeping `dbname` or `database` as optional and then eventually remove it. I fear if we immediately get rid of this, it will  break things for users who are currently using `dbname` or `database`.

**Comment by ThoSap** [comment]
> Thanks for explaining this to me. I understand the following.
> 
> * You would like to keep `dbname` or `database` as optional parameter in connection config. This way it will not be a breaking change for Airbyte. Also, you can create a PR there to use the new dbt adapter
> * Secondly,` dbname` or `database` is redundant in the WHERE clause in `oracle__list_relations_without_caching` macro and we can safely remove it. Thus we can leave it empty or None
> 
> Let me know if this understanding is correct.

- I would like to **remove** `dbname`, `database` and env var `DBT_ORACLE_DATABASE` alltogether as it makes no sense at all. It does not matter if one connects using `OracleConnectionMethod` -> `HOST`, `TNS `or `CONNECTION_STRING`, you will always be on the same database (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) and therefore also the function [impl.py#verify_database](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/impl.py#L101-L112) can _never_ throw an exception, not even with a DB link.

- The WHERE clause in macros
  [`oracle__list_relations_without_caching`](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/include/oracle/macros/adapters.sql#L344)
    and
  [`oracle__get_columns_in_relation`](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/include/oracle/macros/adapters.sql#L185-L187)
    also do not make sense at all due to the first point. 
    We do not have to filter a column just for the sake that it exists in the first place.

Maybe I'm missing the technical reason but why was this `dbname`, `database` and env var `DBT_ORACLE_DATABASE` introduced in the first place?
Could it be that with the Oracle Autonomous Database Service (ADBS) it is possible to have multiple database (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) names with the same service/connection as I never used ADBS before.

**Comment by geoHeil** [comment]
> > Thanks for explaining this to me. I understand the following.
> > 
> > * You would like to keep `dbname` or `database` as optional parameter in connection config. This way it will not be a breaking change for Airbyte. Also, you can create a PR there to use the new dbt adapter
> > * Secondly,` dbname` or `database` is redundant in the WHERE clause in `oracle__list_relations_without_caching` macro and we can safely remove it. Thus we can leave it empty or None
> > 
> > Let me know if this understanding is correct.
> 
>     * I would like to **remove** `dbname`, `database` and env var `DBT_ORACLE_DATABASE` alltogether as it makes no sense at all. It does not matter if one connects using `OracleConnectionMethod` -> `HOST`, `TNS `or `CONNECTION_STRING`, you will always be on the same database (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) and therefore also the function [impl.py#verify_database](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/impl.py#L101-L112) can _never_ throw an exception, not even with a DB link.
> 
>     * The WHERE clause in macros
>       [`oracle__list_relations_without_caching`](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/include/oracle/macros/adapters.sql#L344)
>       and
>       [`oracle__get_columns_in_relation`](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/include/oracle/macros/adapters.sql#L185-L187)
>       also do not make sense at all due to the first point.
>       We do not have to filter a column just for the sake that it exists in the first place.
> 
> 
> Maybe I'm missing the technical reason but why was this `dbname`, `database` and env var `DBT_ORACLE_DATABASE` introduced in the first place? Could it be that with the Oracle Autonomous Database Service (ADBS) it is possible to have multiple database (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) names with the same service/connection as I never used ADBS before.

agree and would love to see this simplification/fix getting merged.

File: dbt/adapters/oracle/impl.py [hunk 1]
```diff
@@ -98,18 +99,13 @@ def convert_number_type(cls, agate_table, col_idx):
     def convert_time_type(cls, agate_table, col_idx):
         return "timestamp"
 
-    @available
-    def verify_database(self, database):
```
Comment (ThoSap) [inline dbt/adapters/oracle/impl.py:102]: This check makes no sense as no matter if one connects using `OracleConnectionMethod` -> `HOST`, `TNS` or `CONNECTION_STRING`, you will always be on the same database (`select SYS_CONTEXT('userenv', 'DB_NAME') from dual;`) and therefore the exception will never be thrown.

**Comment by aosingh** [comment]
> Maybe I'm missing the technical reason but why was this `dbname`, `database` and env var `DBT_ORACLE_DATABASE` introduced in the first place?

This was present in Techindicium's version and we have kept it for compatibility reasons. I do see your point. Even in ADBS we will not have multiple values for `db_name`.

**Comment by aosingh** [comment]
Also, even in dbt-core v0.19 (one used by Airbyte), `database` is a mandatory parameter in the Credentials class. These contracts are defined by `dbt-core`. Maybe this is also why `database` or `dbname` existed in dbt-oracle

https://github.com/dbt-labs/dbt-core/blob/0.19.latest/core/dbt/contracts/connection.py#L120

**Comment by aosingh** [comment]
Currently, as part of this PR, we are making changes at a lot of places just to get one parameter removed. I am evaluating between making so many changes v/s simply specifying the correct value of `database`. You can always get the correct value. It will be helpful if you could help me understand the following things

1. `database` was always a mandatory parameter. How is this a breaking change for Airbyte ? 

2. Based on the code changes why is this the case ? -> `database can be equal to schema`. The user is always the same as schema. Not database. 

For the below error, it has got to do with how inheritance is handled in Python dataclasses

>TypeError: non-default argument 'user' follows default argument

The superclass Credentials has 2 mandatory positional attributes -> `[database, schema]`
As part of `OracleCredentials` we add `[database=None, schema=X, user, pass, ...]` This ordered list of attributes is passed to the `__init__()` method and you get an error because `user` is a positional argument which follows `schema` which is keyword argument

**Review by aosingh — COMMENTED** [review]
This will raise `TypeError`, it has got to do with how inheritance is handled in Python dataclasses

    TypeError: non-default argument 'user' follows default argument

The superclass Credentials has 2 mandatory positional attributes -> [database, schema]
As part of OracleCredentials we add [database=None, schema=X, user, pass, ...] This ordered list of attributes is passed to the __init__() method and you get an error because user is a positional argument which follows schema which is keyword argument

To fix this, an option is to provide default values for every parameter

File: dbt/adapters/oracle/relation.py [hunk 2]
```diff
@@ -39,8 +39,20 @@ class OracleRelation(BaseRelation):
...
         return f'dbt__cte__{name}__'
 
-
+    def render(self):
+        if self.include_policy.database and self.include_policy.schema:
+            raise dbt.exceptions.RuntimeException(
+                'Got a Oracle relation with schema and database set to '
+                'include, but only one can be set'
+            )
+    return super().render()
```
Comment (aosingh) [inline dbt/adapters/oracle/relation.py:58]: This is not syntactically correct. Indentation is missing

**Comment by aosingh** [comment]
> > I just wanted to create an **Individual OCA** but the `oracle/dbt-oracle` repo is not available for selection and it is a required field. https://oca.opensource.oracle.com/api/v1/projects/enabled
> > Can you help @aosingh ?
> 
> Do you have an update on this @aosingh ? I would sign the OCR in the meantime.

I have raised this internally so that dbt-oracle becomes available for selection.

**Comment by ThoSap** [comment]
> Currently, as part of this PR, we are making changes at a lot of places just to get one parameter removed. I am evaluating between making so many changes v/s simply specifying the correct value of `database`. You can always get the correct value. It will be helpful if you could help me understand the following things
> 
> 1. `database` was always a mandatory parameter. How is this a breaking change for Airbyte ?
> 2. Based on the code changes why is this the case ? -> `database can be equal to schema`. The user is always the same as schema. Not database.
> 
> For the below error, it has got to do with how inheritance is handled in Python dataclasses
> 
> > TypeError: non-default argument 'user' follows default argument
> 
> The superclass Credentials has 2 mandatory positional attributes -> `[database, schema]` As part of `OracleCredentials` we add `[database=None, schema=X, user, pass, ...]` This ordered list of attributes is passed to the `__init__()` method and you get an error because `user` is a positional argument which follows `schema` which is keyword argument

You are right, we are making a lot of changes but in my opinion, this major _1.0.0_ release is very fresh and probably not many people are using this release already.
So it would be the **perfect opportunity** to drop this useless parameter altogether as it is internally only used in two where clauses where it is not really needed at all.

**1.** The `database` parameter was never specified in Airbyte because of the following edge case when `dbname` is specified but `service` is not (see quoted example below).
Don't ask me why, I just recently started to use Airbyte and just want to contribute so that the Oracle normalization (JSON to normalized table with columns) works in Airbyte with dbt-oracle 1.0.0.
See the Airbyte code and history https://github.com/airbytehq/airbyte/blob/v0.38.1-alpha/airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py#L250-L264

> This probably has historic reasons and also there is the confusion, where if one specifies the **profiles.yml** property `dbname` **but not** `database` and also omits `service`, the property `dbname` with alias [database](https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/connections.py#L83) gets to be the service name and then the connection works anyways. https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/connections.py#L129-L134 Example **profiles.yml**:
> 
> ```yaml
> config:
>   partial_parse: true
>   printer_width: 120
>   send_anonymous_usage_stats: false
>   use_colors: true
> dbt_test:
>   target: prod
>   outputs:
>     prod:
>       type: oracle
>       protocol: tcp
>       host: yourhost.example.com
>       port: 1521
>       schema: test
>       user: test
>       pass: SECRET_PW
>       # Connection works anyways as in connections.py#L83 the dbname property gets alias database
>       # and if the service property is not specified the database internally is the service in the end
>       # https://github.com/oracle/dbt-oracle/blob/v1.0.0/dbt/adapters/oracle/connections.py#L134
>       dbname: test.service.name.example.com
>       threads: 4
> ```

Also if a user specifies the wrong `database`  property value, let's say the user specifies the database name with a trailing space by accident -> `'mydatabase '` you will get the following error with `dbt-oracle`:
https://github.com/dbt-labs/dbt-core/issues/3470
This error tells the user absolutely nothing about what is going on and why it is not working.

**2.** You are absolutely right, this makes no sense for Oracle.
This was just some copy pasta from `dbt-mysql` and `dbt-clickhouse`, sorry about that 😁

I specified a default value for every property, so the `TypeError: non-default argument 'argument' follows default argument` is not an issue anymore.

**Comment by ThoSap** [comment]
> Thank you @ThoSap for testing and contributing.
> 
> While we review to get this merged, could you sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/) ?
> 
> The bottom of your commit message must have the following line using your name and e-mail address as it appears in the OCA Signatories list.
> 
> ```
> Signed-off-by: Your Name <you@example.org>
> ```
> 
> This can be automatically added to pull requests by committing with:
> 
> ```
> git commit --signoff
> ```

I signed off all commits using my OCA agreement using rebase, sorry for the commit spam here.
https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/changing-a-commit-message
https://hyperledger-indy.readthedocs.io/projects/sdk/en/latest/docs/contributors/signing-commits.html#signing-your-current-commit

**Comment by ThoSap** [comment]
Superseded by https://github.com/oracle/dbt-oracle/pull/6

**Pull request closed without merging**